py/src/nupic/examples/tm/tm_overlapping_sequences.py

```diff
@@ -0,0 +1,917 @@
+    tmStats[name] = stats = tm.getStats()
+    if VERBOSITY >= 2:
+      print("Stats for %s TM over all sequences:" % (name))
+      print("   total missing:", stats['totalMissing'])
+      print("   total extra:", stats['totalExtra'])
+
+  for (name,tm) in tms.items():
+    if VERBOSITY >= 3:
+      print("\nAll internal stats for %s TM:" % (name))
+      print("-------------------------------------", end=' ')
+      pprint.pprint(tmStats[name])
+      print()
+
+  return tmStats
+
+
+
+def testConfig(baseParams, expMissingMin=0, expMissingMax=0, **mods):
+  """
+  Build up a set of sequences, create the TM(s), train them, test them,
+  and check that we got the expected number of missing predictions during
+  inference.
+
+  Parameters:
+  -----------------------------------------------------------------------
+  baseParams:     dict of all of the parameters for building sequences,
+                      creating the TMs, and training and testing them. This
+                      gets updated from 'mods' before we use it.
+
+  expMissingMin:   Minimum number of expected missing predictions during testing.
+  expMissingMax:   Maximum number of expected missing predictions during testing.
+
+  mods:           dict of modifications to make to the baseParams.
+  """
+
+
+  # Update the base with the modifications
+  params = dict(baseParams)
+  params.update(mods)
+
+  # --------------------------------------------------------------------
+  # Create the sequences
+  func = params['seqFunction']
+  (numCols, trainingSequences) = func(**params)
+
+  # --------------------------------------------------------------------
+  # Create the TMs
+  if params['numCols'] is None:
+    params['numCols'] = numCols
+  tms = createTMs(**params)
+
+  # --------------------------------------------------------------------
+  # Train and get test results
+  tmStats = evalSequences(tms= tms,
+                          trainingSequences=trainingSequences,
+                          testSequences=None,
+                          **params)
+
+  # -----------------------------------------------------------------------
+  # Make sure there are the expected number of missing predictions
+  for (name, stats) in tmStats.items():
+    print("Detected %d missing predictions overall during inference" \
+              % (stats['totalMissing']))
+    if expMissingMin is not None and stats['totalMissing'] < expMissingMin:
+      print("FAILURE: Expected at least %d total missing but got %d" \
+          % (expMissingMin, stats['totalMissing']))
+      assert False
+    if expMissingMax is not None and stats['totalMissing'] > expMissingMax:
+      print("FAILURE: Expected at most %d total missing but got %d" \
+          % (expMissingMax, stats['totalMissing']))
+      assert False
+
+
+  return True
+
+
+class TMOverlappingSeqsTest(testcasebase.TestCaseBase):
+
+  def testFastLearning(self):
+    """
+    Test with fast learning, make sure PAM allows us to train with fewer
```

Comment:
What is PAM?   I assume P is probably `Prediction` but I don't know what 'A' and 'M' are.

Reply:
PAM stands for "Pay Attention Mode". It was part of the backtracking TM.

Reply:
Ah.  I guess I should have remembered that.